src/main/java/org/openrewrite/java/spring/boot2/upgrade/to25/FindClassesDependingOnDataSource.java

```diff
@@ -0,0 +1,116 @@
+
+        public void add(List<String> classDeclarations) {
+            this.classDeclartions.addAll(classDeclarations);
+        }
+
+        public boolean contains(String classDecl) {
+            return classDeclartions.contains(classDecl);
+        }
+
+        public List<String> getAll() {
+            return Collections.unmodifiableList(classDeclartions);
+        }
+    }
+
+    @Override
+    protected JavaIsoVisitor<ExecutionContext> getVisitor() {
+        return new JavaIsoVisitor<ExecutionContext>() {
+
+            @Override
+            public J.CompilationUnit visitCompilationUnit(J.CompilationUnit compilationUnit, ExecutionContext executionContext) {
+                J.CompilationUnit cu = super.visitCompilationUnit(compilationUnit, executionContext);
+
+                if (executionContext.getMessage(CLASSES_USING_DATA_SOURCE) == null) {
+                    Matches classDeclsDependingOnDataSource = new Matches();
+                    executionContext.putMessage(CLASSES_USING_DATA_SOURCE, classDeclsDependingOnDataSource);
+                }
+
+                if(isNotProcessed(cu, executionContext) && hasImportsOfTypeDataSource(cu)) {
+                    List<J.ClassDeclaration> matchingClassDecls = findClassDeclarationsDependingOnDataSource(cu, executionContext);
+                    if( ! matchingClassDecls.isEmpty()) {
+                        storeClassDeclInExecutionContext(matchingClassDecls, executionContext);
+                    }
+                    return cu;
+                }
+
+                return cu;
+            }
+
+            private boolean isNotProcessed(J.CompilationUnit cu, ExecutionContext executionContext) {
+                return cu.getClasses().stream().noneMatch(c -> FindClassesDependingOnDataSource.isMatch(c, executionContext));
+            }
+
+            private void storeClassDeclInExecutionContext(List<J.ClassDeclaration> classDeclarations, ExecutionContext executionContext) {
+                Matches messages = executionContext.getMessage(CLASSES_USING_DATA_SOURCE);
+                messages.add(classDeclarations.stream().map(cd -> cd.getType().getFullyQualifiedName()).collect(Collectors.toList()));
+            }
+
+
+            private List<J.ClassDeclaration> findClassDeclarationsDependingOnDataSource(J.CompilationUnit cu, ExecutionContext executionContext) {
+                return cu.getClasses().stream()
+                        .filter(cd -> typesDependingOnDataSource(cd, executionContext, cu))
+                        .collect(Collectors.toList());
+            }
+
+            private boolean typesDependingOnDataSource(J.ClassDeclaration classDeclaration, ExecutionContext executionContext, J.CompilationUnit cu) {
+                // FIXME: does not recognize dependency on DataSource
+                J.ClassDeclaration classDeclaration1 = (J.ClassDeclaration) new UsesType<>(JAVAX_DATA_SOURCE).visitClassDeclaration(classDeclaration, executionContext);
```

Comment:
@pway99 What was the approach you mentioned yesterday to find out if a `ClassDeclaration` depends on a certain type?

Reply:
`FindTypes.findAssignable(classDeclaration, JAVAX_DATA_SOURCE)` will return an empty list if the ClassDeclaration does not have any instances JAVAX_DATA_SOURCE

Reply:
Nice, will try.
I also tried to define the finder as internal class of `InitializationOrdering` and move the visitor from the `AnnotateClassesDependingOnDataSource` to `InitializationOrdering`.
Then using the find visitor in the visit(before,..) method. But as the finder does not add markers anymore the result is empty (didn't try though). If the finder visitor would provide a result it would work and we'd have only one (huge) class with the search visitor as internal class that can ve accessed from outside...

Reply:
what if the finder was a nested class something like:
```java
private static class FindClassesHavingDataSource extends JavaIsoVisitor<ExecutionContext> {
    private final List<String> classesHavingDataSource;

    public FindDsClasses(List<String> classesHavingDataSource) {
        this.classesHavingDataSource = classesHavingDataSource;
    }

    @Override
    public J.CompilationUnit visitCompilationUnit(J.CompilationUnit compilationUnit, ExecutionContext executionContext) {
        J.CompilationUnit cu = super.visitCompilationUnit(compilationUnit, executionContext);
        for (J.ClassDeclaration cd : cu.getClasses()) {
            if (!FindTypes.findAssignable(cd, JAVAX_DATA_SOURCE).isEmpty()) {
                classesHavingDataSource.add(cd.getType().getFullyQualifiedName());
            }
        }
        return cu;
    }
}
```

Reply:
What difference would this make?

Reply:
It could be used for adding classes having DataSource references to the execution context. something like:
```java
@Override
protected List<SourceFile> visit(List<SourceFile> before, ExecutionContext ctx) {
    List<String> dsClasses = new ArrayList<>();
    FindClassesHavingDataSource finder = new FindClassesHavingDataSource(dsClasses);
    for (SourceFile sf : before) {
        finder.visit(sf, ctx);
    }
    ctx.putMessage("DS_CLASSES", dsClasses);
    return super.visit(before, ctx);
}
```